File: Data Structures & Algorithms/longest-substring-without-duplicates/submission-5.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int l = 0;
        int longest = 0;
        Set<Character> seen = new HashSet<>();

        for (int r = 0; r < s.length(); r++ ) {
            while (seen.contains(s.charAt(r))) {
                seen.remove(s.charAt(l++));
            }

            seen.add(s.charAt(r));
            longest = Math.max(longest, r - l + 1);
        }
        
        return longest;
    }
}

File: Data Structures & Algorithms/buy-and-sell-crypto/submission-5.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        
        l, r, profit = 0, 0, 0

        while r < len(prices):

            while l < r and prices[l] > prices[r]:
                l += 1

            profit = max(profit, prices[r] - prices[l])
            r += 1

        return profit 